File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n=len(s1)
        m=len(s2)
        if m<n:
            print("False")
            return False
        a=[0]*26
        b=[0]*26
        for i in range(n):
            a[ord(s1[i])-ord("a")]+=1
            b[ord(s2[i])-ord("a")]+=1 

        for i in range(m-n):
            if a==b:
                return True
            b[ord(s2[i])-ord("a")]-=1
            b[ord(s2[i+len(s1)])-ord("a")]+=1

        return a==b

File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int n=s1.length();
        int m=s2.length();
        int i;
        if(n>m){
            return false;
        }
        int[] a = new int[26];
        int[] b = new int[26];
        for(i=0;i<n;i++){
            a[s1.charAt(i)-'a']++;
            b[s2.charAt(i)-'a']++;
        }
        for (i = 0; i <= m - n; i++) {
            if (Equal(a, b)) {  
                return true; 
            }
            if (i < m - n) {
                b[s2.charAt(i) - 'a']--; 
                b[s2.charAt(i + n) - 'a']++;
            }
        }
        return Equal(a, b);
    }
    private static boolean Equal(int[] a, int[] b) {
        for (int i = 0; i < 26; i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }
        return true;
    }
    }